server/src/main/java/io/crate/execution/dml/RawIndexer.java

```diff
@@ -92,26 +100,54 @@ public ParsedDocument index(IndexItem item) throws IOException {
                 returnValues
             );
         });
-        Object[] insertValues = new Object[doc.size()];
+
+        currentRowValues = new Object[doc.size()];
         Iterator<Object> iterator = doc.values().iterator();
-        List<Reference> columns = indexer.columns();
-        for (int i = 0; i < insertValues.length; i++) {
+        List<Reference> columns = currentRowIndexer.columns();
+        for (int i = 0; i < currentRowValues.length; i++) {
             Reference reference = columns.get(i);
             Object value = iterator.next();
             DataType<?> type = reference.valueType();
             try {
-                insertValues[i] = type.implicitCast(value);
+                currentRowValues[i] = type.implicitCast(value);
             } catch (ClassCastException | IllegalArgumentException e) {
                 throw new ConversionException(value, type);
             }
         }
-        ParsedDocument parsedDoc = indexer.index(new IndexItem.StaticItem(
+
+        return currentRowIndexer.collectSchemaUpdates(new IndexItem.StaticItem(
             item.id(),
             item.pkValues(),
-            insertValues,
+            currentRowValues,
+            item.seqNo(),
+            item.primaryTerm()
+        ));
+    }
+
+    /**
+     * Create a {@link ParsedDocument} from {@link IndexItem}
+     *
+     * This must be called after any new columns (found via
+     * {@link #collectSchemaUpdates(IndexItem)}) have been added to the cluster
+     * state.
+     */
+    public ParsedDocument index(IndexItem item) throws IOException {
+        assert currentRowIndexer != null : "Must be used only after collecting schema updates.";
+
+
+        // Replica must use the same values for undeterministic defaults/generated columns
+        if (currentRowIndexer.hasUndeterministicSynthetics()) {
+            currentRowValues = currentRowIndexer.addGeneratedValues(currentRowValues);
+        }
```

Comment:
We BTW don't have such check on master, probably bug?

I was actually fixing `test_generated_non_deterministic_value_is_consistent_on_primary_and_replica` but then realized it has nothing to do with COPY FROM. 


So we probably need similar test case for COPY FROM?